File: December2022/Java/HouseRobber.java
// @saorav21994

class Solution {
    public int rob(int[] nums) {
        int l = nums.length;
        if (l == 1)
            return nums[0];
        int dp2 = nums[0];
        int dp1 = Math.max(nums[0], nums[1]);
        int dp = dp1;
        for (int i = 2; i < l; i++) {
            dp = Math.max(dp2 + nums[i], dp1);
            dp2 = dp1;
            dp1 = dp;
        }
        return dp;
    }
}

File: December2022/C++/house-robber.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        
        int numHouses = nums.size();
        
        if(numHouses ==1){
            return nums[0];
        }
        
        int prevHouse1 = nums[0];
        

        
        int ans = max(nums[0], nums[1]);
                int prevHouse2 = ans;
        for(int house =2; house < numHouses; house++){
            
            ans = max( nums[house] + prevHouse1,prevHouse2);
            prevHouse1 = prevHouse2;
            prevHouse2 = ans;
            
        }
        
        return ans;
    }
};